File: src/main/java/br/com/jiratorio/service/impl/TokenServiceImpl.java
package br.com.jiratorio.service.impl;

import br.com.jiratorio.domain.vo.Account;
import br.com.jiratorio.exception.ParseTokenException;
import br.com.jiratorio.service.TokenService;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.jwt.Jwt;
import org.springframework.security.jwt.JwtHelper;
import org.springframework.security.jwt.crypto.sign.SignatureVerifier;
import org.springframework.security.jwt.crypto.sign.Signer;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor(onConstructor = @__(@Autowired))
public class TokenServiceImpl implements TokenService {

    private final ObjectMapper objectMapper;

    private final SignatureVerifier signatureVerifier;

    private final Signer signer;

    @Override
    public String encode(final Account account) {
        log.info("Method=encode, account={}", account);
        try {
            String json = objectMapper.writeValueAsString(account);
            Jwt jwt = JwtHelper.encode(json, signer);
            return jwt.getEncoded();
        } catch (Exception e) {
            throw new ParseTokenException(e);
        }
    }

    @Override
    public Account decode(final String token) {
        log.info("Method=decode, token={}", token);

        try {
            Jwt jwt = JwtHelper.decodeAndVerify(token, signatureVerifier);
            return objectMapper.readValue(jwt.getClaims(), Account.class);
        } catch (Exception e) {
            throw new ParseTokenException(e);
        }
    }
}

File: src/main/java/br/com/jiratorio/service/impl/TokenServiceImpl.kt
package br.com.jiratorio.service.impl

import br.com.jiratorio.domain.Account
import br.com.jiratorio.service.TokenService
import br.com.jiratorio.extension.logger
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.security.jwt.JwtHelper
import org.springframework.security.jwt.crypto.sign.SignatureVerifier
import org.springframework.security.jwt.crypto.sign.Signer
import org.springframework.stereotype.Service

@Service
class TokenServiceImpl(
    private val objectMapper: ObjectMapper,
    private val signatureVerifier: SignatureVerifier,
    private val signer: Signer
) : TokenService {

    private val log = logger()

    override fun encode(account: Account): String {
        log.info("Method=encode, account={}", account)

        val json = objectMapper.writeValueAsString(account)
        return JwtHelper.encode(json, signer).encoded
    }

    override fun decode(token: String): Account {
        log.info("Method=decode, token={}", token)

        val jwt = JwtHelper.decodeAndVerify(token, signatureVerifier)
        return objectMapper.readValue(jwt.claims, Account::class.java)
    }
}
